arelle/ModelXbrl.py

```diff
@@ -771,51 +671,49 @@ def factsByQname(self): # indexed by fact (concept) qname
             return fbqn
 
     @property
-    def factsByLocalName(self): # indexed by fact (concept) localName
+    def factsByLocalName(self) -> dict[QName, set[ModelInlineFact]]:  # indexed by fact (concept) localName
         """Facts in the instance indexed by their LocalName, cached
-
-        :returns: dict -- indexes are LocalNames, values are ModelFacts
         """
+        self._factsByLocalName: dict[QName, set[ModelInlineFact]]
         try:
             return self._factsByLocalName
         except AttributeError:
+            fbln: dict[QName, set[ModelInlineFact]]
             self._factsByLocalName = fbln = defaultdict(set)
             for f in self.factsInInstance:
                 if f.qname is not None:
                     fbln[f.qname.localName].add(f)
             return fbln
 
-    def factsByDatatype(self, notStrict, typeQname): # indexed by fact (concept) qname
+    def factsByDatatype(self, notStrict: bool, typeQname: QName) -> set[ModelFact] | None:  # indexed by fact (concept) qname
         """Facts in the instance indexed by data type QName, cached as types are requested
 
         :param notSctrict: if True, fact may be derived
-        :type notStrict: bool
-        :returns: set -- ModelFacts that have specified type or (if nonStrict) derived from specified type
         """
         try:
             return self._factsByDatatype[notStrict, typeQname]
         except AttributeError:
             self._factsByDatatype = {}
             return self.factsByDatatype(notStrict, typeQname)
         except KeyError:
+            fbdt: set[ModelFact]
             self._factsByDatatype[notStrict, typeQname] = fbdt = set()
             for f in self.factsInInstance:
                 c = f.concept
                 if c.typeQname == typeQname or (notStrict and c.type.isDerivedFrom(typeQname)):
                     fbdt.add(f)
             return fbdt
 
-    def factsByPeriodType(self, periodType): # indexed by fact (concept) qname
+    def factsByPeriodType(self, periodType: str) -> Set[ModelFact]:  # indexed by fact (concept) qname
         """Facts in the instance indexed by periodType, cached
 
         :param periodType: Period type to match ("instant", "duration", or "forever")
-        :type periodType: str
-        :returns: set -- ModelFacts that have specified periodType
         """
         try:
             return self._factsByPeriodType[periodType]
         except AttributeError:
-            self._factsByPeriodType = fbpt = defaultdict(set)
+            self._factsByPeriodType: defaultdict[str, set[ModelFact]] = defaultdict(set)
+            fbpt: defaultdict[str,set[ModelFact]] = defaultdict(set)
```

Comment:
This is a breaking change. `self._factsByPeriodType` and the local `fbpt` are no longer the same var. You're populating `fbpt`, but returning values from `self._factsByPeriodType`.